Creating CompositePropagator

eddynaka opened the pull request:
Fixes #582.

## Changes
- Changed the `ITextFormat.Extract` to accept `ActivityContext` based on the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#extract
- Added `CompositePropagator` based on the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#create-a-composite-propagator
  - Added tests
  - removed duplicated (`TraceContextActivityTest`)

Thanks to @CodeBlanch and @cijothomas for the guidance!

Please provide a brief description of the changes here. Update the
`CHANGELOG.md` for non-trivial changes.

For significant contributions please make sure you have completed the following items:

* [ ] Design discussion issue #
* [ ] Changes in public API reviewed

CodeBlanch commented on src/OpenTelemetry.Api/Context/Propagation/CompositePropagator.cs:
```diff
@@ -0,0 +1,68 @@
+        /// <inheritdoc/>
+        public bool IsInjected<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        {
+            throw new NotImplementedException();
```
Pretty easy to implement this one. Just do:
```csharp
public bool IsInjected<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
{
            foreach (var textFormat in this.textFormats)
            {
                if (textFormat.IsInjected(carrier, getter))
                {
                     return true;
                }
            }
            return false;
}
```
Without the implementation, users won't be able to plug this in on the HttpClient instrumentation.

This ties into how we have `Extract` working now. As long as one of the child `ITextFormat` has a valid context, it will be extracted.

eddynaka replied:
@CodeBlanch , what if: the first one is injected and the second one not. We would return true. Would that be right?

eddynaka replied:
what i would think is: if one of the textformats return false, so, false, otherwise, its true. in that case, if the last one is false, it would inject everything again.

CodeBlanch replied:
@eddynaka That works for me (return true only if all are injected). It isn't part of the spec so it's up to us to define it for now.

eddynaka replied:
just updated and added some tests to cover both cases

CodeBlanch commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,146 @@
+            Assert.Equal(expected, carrier);
+
+            bool isInjected = compositePropagator.IsInjected(carrier, Getter);
+            Assert.True(isInjected);
```
It feels kind of strange that it returns `true` for `IsInjected` when empty. If `IsInjected` resolves `true` you would expect `Extract` would return a valid context right? Not sure what to do about it. Kind of second-guessing the `IsInjected` API. I would say let's remove it if we didn't need it for the http instrumentation. We could have `ShouldInject` and `CanExtract` but that is heavy for what is essentially an edge case?

eddynaka replied:
since is just for this edge case, i would leave this for now, so we get compliant with the spec and think about what would be the best for us, since that isn't specified in the spec and, we could ask the community what they think.

MikeGoldsmith replied:
Yeah, I agree it's odd that we return true when the text format list is empty. Maybe we could/should return false if the list of formats is empty?

eddynaka replied:
i'm updating :)

cijothomas commented on src/OpenTelemetry.Api/Context/Propagation/ITextFormat.cs:
```diff
@@ -45,10 +45,11 @@ public interface ITextFormat
         /// Extracts activity context from textual representation.
         /// </summary>
         /// <typeparam name="T">Type of object to extract context from. Typically HttpRequest or similar.</typeparam>
+        /// <param name="activityContext">Activity context to transmit over the wire.</param>
```
comment is not so correct.. Extract is to "extract" context from the wire.

eddynaka replied:
good catch! i will update!

cijothomas commented on src/OpenTelemetry.Api/Context/Propagation/TraceContextFormat.cs:
```diff
@@ -74,18 +74,18 @@ public bool IsInjected<T>(T carrier, Func<T, string, IEnumerable<string>> getter
 
         /// <inheritdoc/>
-        public ActivityContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        public ActivityContext Extract<T>(ActivityContext activityContext, T carrier, Func<T, string, IEnumerable<string>> getter)
```
Not sure i understood this.. An existing context is passed.. We extract and create a new context and return it.. so why would we need the existing context? Its currently used as returned context when nothing can be extracted.. I was imagining more like - we build on the given activity context.. So the returned one has all traceid,spanid,tracestate it extracted, along with whatever tracestate already present in the passed activitycontext.

I want to hear expert thoughts here @reyang @CodeBlanch

eddynaka replied:
For other classes that implement the ITextFormat, the activityContext passed would be the default, so, if something happen, for example, the carrier is null or other validation, we would return default. So, nothing changed.

For the case where we have the CompositePropagator, so, the first time, that would be the default, and, from that to the next one, it will keep passing and reutilizing the activitycontext. With that, the carrier would have in the end all the data from all the ITextFormat.

Below, the link from OT.python: https://github.com/open-telemetry/opentelemetry-python/blob/004896f1987e84b2d8f5c84bb1505099e178694c/opentelemetry-api/src/opentelemetry/propagators/composite.py#L23
Below, the link from OT.java: https://github.com/open-telemetry/opentelemetry-java/blob/690651b65a938c903c35d65aef3182999fcb1b19/context_prop/src/main/java/io/opentelemetry/context/propagation/DefaultContextPropagators.java#L142-L147

cijothomas replied:
If
SystemA extractor gives me context with IDs and TraceState as "systemAField=something"
SystemB extractor gives me context with IDs and TraceState as "systemBField=somethingelse" 

The if I use a composite propagator cosisting of "SystemA", "SystemB", should I get the context with IDs from SystemB (as its added after A?), and TraceState as combining both "systemAField=something,systemBField=somethingelse"

Is the above correct? The spec seems vague about this. Hopefully Reiley/Mikel have more context on context.

CodeBlanch replied:
This was a tough one! When I first looked at @eddynaka's implementation, it was like this:

```csharp
        public ActivityContext Extract<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
        {
            foreach (var textFormat in this.textFormats)
            {
                activityContext = textFormat.Extract(carrier, getter);
            }

            return activityContext;
        }
```

Loop through the child propagators, extract from them. But what if the first one succeeds, returns a valid context, and then all others fail? You end up with no context.

Initially, we thought the spec was bugged. But looking at py and java, we realized we were missing the `context` parameter on the `Extract` "base" operation: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#extract

The current version is actually how the spec says to do it, and how py and java are doing it! It seems OK to me. The goal is to extract the context. What `CompositePropagator` is doing is allowing to you inject in multiple formats when you transmit. Presumably to support mixing in with multiple downstream systems? When you extract it, you really only need one version of the context. All the others should be duplicates, so the loop really could return on the first valid one it finds?

But agree the spec is a bit strange and vague definitely for propagation. We should try to get it improved, and hammer out how the `IsInjected` should work (https://github.com/open-telemetry/opentelemetry-specification/issues/674).

cijothomas replied:
674 got tagged for after-ga, it won't likely be resolved this year :-:(

So if the 1st Extract is a custompropogator which looked at "customhttpheader" and extract a Context, we return that right? If that custompropogator is not extracting the W3C `tracestate` header, then we are likely losing that information unless every extractor is given a chance to run and cumulatively build `tracestate`. 

Also even if any number of extractors are run, they should either not be able to obtain traceid/spanid, or obtain traceid/span which would match others extractors who are able to successfully extract. -- is this correct?

If I am wrong in my tracestate interpretation and propagation, I'd love to get it clarified. Hopefully then then spec can be made more clear. Today spec simply says , for "tracestate" , refer to W3C spec.

MikeGoldsmith replied:
Hello :wave:

When using a composite propagator, the first successfully extracted context should be returned, ignoring the remaining propagators (if any). This is to allow contexts to be extracted from different formats, in a multiformat setup but, as @CodeBlanch said, if multiple formats are received, they should contain the same information.

eddynaka replied:
@MikeGoldsmith , based on that, should we apply the same strategy to IsInjected? If we have one that is injected, so, return true?

eddynaka replied:
just updated. Let me know what do u think

cijothomas replied:
After reading more times, I come to think that this is correct.

CodeBlanch commented on src/OpenTelemetry.Api/Context/Propagation/CompositePropagator.cs:
```diff
@@ -0,0 +1,76 @@
+        {
+            foreach (var textFormat in this.textFormats)
+            {
+                activityContext = textFormat.Extract(activityContext, carrier, getter);
```
@eddynaka Might be able to return on the first valid context found here. Question is, do we need to execute them all just in case they mutate state on carrier?

eddynaka replied:
Should we open a new issue on the spec asking for clarification?

eddynaka replied:
based on my last change, if we find one valid, we would return it.

MikeGoldsmith commented on src/OpenTelemetry.Api/Context/Propagation/CompositePropagator.cs:
```diff
@@ -0,0 +1,76 @@
+        }
+
+        /// <inheritdoc/>
+        public ISet<string> Fields => throw new NotImplementedException();
```
I think it's dangerous to throw here because otherwise you're expecting any code that wants to get the Fields to do the check on this implementation to see if it's a `CompositePropagtor`. I think we should just return an empty set instead.

eddynaka replied:
just updated

cijothomas commented on src/OpenTelemetry.Api/Context/Propagation/ITextFormat.cs:
```diff
@@ -45,10 +45,11 @@ public interface ITextFormat
         /// Extracts activity context from textual representation.
         /// </summary>
         /// <typeparam name="T">Type of object to extract context from. Typically HttpRequest or similar.</typeparam>
+        /// <param name="activityContext">Activity context to extract from the wire.</param>
```
this comment is incorrect, isn't it?

eddynaka replied:
it is. but i don't know how would be the better words to replace. Do you have any suggestion?

MikeGoldsmith replied:
How about: "The default activity context to be used if Extract fails."

cijothomas commented on src/OpenTelemetry.Api/Context/Propagation/CompositePropagator.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <param name="textFormats">List of <see cref="ITextFormat"/> wire context propagator.</param>
+        public CompositePropagator(List<ITextFormat> textFormats = null)
+        {
+            this.textFormats = textFormats ?? new List<ITextFormat>();
```
what is the use case for having empty list?

eddynaka replied:
The idea of this is: if we receive a null list, the code would act the same as if we receive a list with at least one element.
Question:
- should we maintain this?
- should we change to `if list null or empty throw exception`?

cijothomas replied:
If the list is empty, I think we should throw. - Reiley made suggestion about the overall principles  - its okay to throw at initialization stage if bad input is received. But we shouldn't ever throw after startup.

List with '0' element - no need of throwing. But no injection/extraction occurs.

eddynaka replied:
just updated!

cijothomas commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,146 @@
+        }
+
+        [Fact]
+        public void CompositePropagator_WithTraceContext()
```
Composite propagator doesn't care what its composed of - to test it, we can write test propagators, and no use b3 or tracecontext.

eddynaka replied:
I created two new tests:
1. using a custompropagator + TraceContextFormat to check the injection and the extraction, the case where the CustomPropagator doesn't generate a valid ActivityContext and, the second one does
2. using CustomPropagator1 + CustomPropagator2 to check what would happen if both propagators sets the same variable.

eddynaka replied:
just updated!

eddynaka commented:
> My main concern is alleviated. I'd love to see more unit tests validating the behavior in all scenarios.
> Also, this is a breaking change from beta - could you update changelog with this info?

Sure! just updated and added a reference to spec

CodeBlanch commented on src/OpenTelemetry.Api/Trace/SpanContext.cs:
```diff
@@ -121,6 +121,12 @@ public bool IsRemote
+        /// Converts a <see cref="SpanContext"/> into an <see cref="ActivityContext"/>.
+        /// </summary>
+        /// <param name="spanContext"><see cref="SpanContext"/> source.</param>
+        public static implicit operator ActivityContext(SpanContext spanContext) => spanContext.ActivityContext;
```
@cijothomas I'm thinking about dropping implicit casts on all the shim classes. Why do that? Not sure. It just feels like a good thing to have. Could reduce bugs in places where we don't consider the Span case because we focus more on the Activity case? The idea is to allow Span version to be passed anywhere Activity version can be passed. How do you feel about that? 🤷

CodeBlanch commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CustomPropagator1.cs:
```diff
@@ -0,0 +1,45 @@
+
+namespace OpenTelemetry.Tests.Implementation.Trace.Propagation
+{
+    public class CustomPropagator1 : ITextFormat
```
`CustomPropagator1` & `CustomPropagator2` are a bit different than what I was expecting. Why not just have a single `TestPropagator` that you can pass in the header name that it will use for propagation? Should allow you to model everything you need for test, and fully implement the API (like Fields).

cijothomas replied:
@eddynaka Can you make this suggestion?

eddynaka replied:
Hi @cijothomas , i'm talking with @CodeBlanch , to understand the idea.

eddynaka replied:
@cijothomas , can you check again? @CodeBlanch helped me with the refactor.

cijothomas commented on src/OpenTelemetry.Api/CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Added `CompositePropagator` that accepts a list of `ITextFormat` following
```
This is good. 
We also need an entry to say the ITextFormat API has changed. Thats more crucial as its a breaking change from previous beta.

eddynaka replied:
just updated, and included the new method that we will have to implement

MikeGoldsmith commented on src/OpenTelemetry.Api/Context/Propagation/CompositePropagator.cs:
```diff
@@ -0,0 +1,86 @@
+        }
+
+        /// <inheritdoc/>
+        public ISet<string> Fields => new HashSet<string>();
```
I think we should add a `private static readonly ISet<string> EmptyFields = new HastSet<string>();` to the top of this class and use it here.

eddynaka replied:
thank you. just updated

MikeGoldsmith commented on src/OpenTelemetry.Api/Context/Propagation/CompositePropagator.cs:
```diff
@@ -0,0 +1,86 @@
+        /// <inheritdoc/>
+        public bool IsInjected<T>(T carrier, Func<T, string, IEnumerable<string>> getter)
+        {
+            if (!this.textFormats.Any())
```
A super condensed version of this function would be to use the `.All()` LINQ function.

```csharp
return this.TextFormats.All(textFormat => textFormat.IsInjected(carrier, getter));
```

eddynaka replied:
thank you. just updated

MikeGoldsmith commented on src/OpenTelemetry.Api/Context/Propagation/TraceContextFormat.cs:
```diff
@@ -304,5 +283,25 @@ private bool TryExtractTracestate(string[] tracestateCollection, out string trac
+        {
+            if ((c >= '0') && (c <= '9'))
+            {
+                return (byte)(c - '0');
```
`return Convert.ToByte(c)` would work better here than doing inline subtractions / additions.

Same for upper and lower casings below.

eddynaka replied:
Can you validate?

cijothomas commented on src/OpenTelemetry.Api/Context/Propagation/TraceContextFormat.cs:
```diff
@@ -157,7 +156,7 @@ public void Inject<T>(ActivityContext activityContext, T carrier, Action<T, stri
         }
 
-        private bool TryExtractTraceparent(string traceparent, out ActivityTraceId traceId, out ActivitySpanId spanId, out ActivityTraceFlags traceOptions)
+        internal static bool TryExtractTraceparent(string traceparent, out ActivityTraceId traceId, out ActivitySpanId spanId, out ActivityTraceFlags traceOptions)
```
Thanks to @CodeBlanch we'll get better help from .NET Activity API to do the parsing etc in the Rc1 release. We can refactor some of the code here at that time.

cijothomas commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,193 @@
+            var carrier = new Dictionary<string, string>();
+
+            compositePropagator.Inject(activityContext, carrier, Setter);
+            Assert.Contains(carrier, kv => kv.Key == "custom-traceparent");
```
what about the value?

eddynaka replied:
will update soon

cijothomas commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,193 @@
+
+            var compositePropagator = new CompositePropagator(new List<ITextFormat>
+            {
+                new TraceContextFormat(),
```
should we repeat this test with the order reversed?

eddynaka replied:
sure. let me change.

cijothomas commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,193 @@
+            Assert.Contains(carrier, kv => kv.Key == "custom-traceparent");
+
+            bool isInjected = compositePropagator.IsInjected(carrier, Getter);
+            Assert.True(isInjected);
```
Lets validate that while injecting with 2 propagators using same key, the last one wins? There is no requirement that the 2nd one checks that header already exist and backs off right?

Same with extract - validate that we return when we find the 1st valid context.

eddynaka replied:
For the inject part: ok, i did some changes that will validate the value and check if the latest one is the one that we have.
For the isInjected part: all will have to be injected. Since i did one with same header and one with different ones. i think we are covered.
For the extract part: changed the constructor to pass a boolean, so we can define if we will return a default or not. With that in mind, I counted how many times we used the Getter. So, if the first one returns default, we have to access only two times.

cijothomas commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,242 @@
+
+            compositePropagator.Inject(activityContext, carrier, Setter);
+            Assert.Contains(carrier, kv => kv.Key == "custom-traceparent-1");
+            Assert.Contains(carrier, kv => kv.Key == "custom-traceparent-2");
```
Since carrier is a dictionary, how is this possible to have duplicate keys?

cijothomas replied:
Ignore. Didn't see keys are different

cijothomas commented on test/OpenTelemetry.Tests/Implementation/Trace/Propagation/CompositePropagatorTest.cs:
```diff
@@ -0,0 +1,242 @@
+        }
+
+        [Fact]
+        public void CompositePropagator_WithTraceContext()
```
I'd remove all TraceState or B3 test from compositepropagator tests. They are tested elsewhere. CompositePropagator doesn't care what its composed of - it just calls inject/extract on them, and the tests must just validate that behavior.

eddynaka replied:
ok. im removing everything related to B3 and traceContextFormat.

cijothomas commented:
@MikeGoldsmith I merged this to make progress - please let us know if your comments are not addressed. We can follow up. Thanks!
